### PR TITLE
Improve routes and logging

### LINKED ISSUE #59: provide explicit `/update` route
At the moment the update script is the last module that is directly requested by clients.
I would propose to forbid any request to `/modules` by offering a explicit route to handle
package uploads `/update` or `/publish` would then forward to that endpoint.

This change requires only two small changes to controller.xql but we would need to discuss if this is possible without breaking other clients like packagemanager instances.

### PR DESCRIPTION
**Pull request by line-o**
- Having a base-url set in the page template head, allows `templates/page.html` to be used for all pages.
- Drop the html extension from all routes
- Extract the logging functionality into a separate module

closes #59

### CODE REVIEW

File: modules/config.xqm [hunk 1]
```diff
@@ -12,6 +12,10 @@ declare namespace system="http://exist-db.org/xquery/system";
 declare namespace expath="http://expath.org/ns/pkg";
 declare namespace repo="http://exist-db.org/xquery/repo";
 
+
+declare variable $config:login-domain := "org.exist.public-repo.login";
```
Comment (joewiz) [inline modules/config.xqm:16]: How is this related to the PR?

File: modules/config.xqm [hunk 2]
```diff
@@ -12,6 +12,10 @@ declare namespace system="http://exist-db.org/xquery/system";
...
+declare variable $config:base-url := "/exist/apps/public-repo/";
```
Comment (joewiz) [inline modules/config.xqm:17]: This assumes a lot about the server and app, and should be composed of other constants.

**Review by joewiz — APPROVED** [review]
LGTM and achieves the desired goal of trimming excess templates.

File: modules/config.xqm (hunk 1 above)
Comment (line-o) [reply modules/config.xqm:16]: This will make more sense with the next commits

File: modules/config.xqm (hunk 2 above)
Comment (line-o) [reply modules/config.xqm:17]: Well, yes. I thought it is at least better to have it here among the configurations. But in the end this is likely a function that checks for is-proxied and the context as well.

Comment (line-o) [reply modules/config.xqm:17]: The reason I went for a simple string is that the repo will in almost all cases have exactly one URL it will be requested from.

File: modules/config.xqm (hunk 1 above)
Comment (line-o) [reply modules/config.xqm:16]: $config:login-domain is now used in the controller and put-package.xq

File: modules/config.xqm (hunk 2 above)
Comment (line-o) [reply modules/config.xqm:17]: If you have a suggestion which variables can be used to construct the base-url more reliably I am all ears!

Comment (joewiz) [reply modules/config.xqm:17]: - `/exist` is available via `request:get-context-path()`. 
- `/apps` is available via `$exist:prefix`
- `public-repo` is available via `$exist:controller`

In other words, the components hardcoded here in `$config:base-url` are actually dependent 3 system variables: [the Jetty webapp context](https://github.com/eXist-db/exist/blob/develop/exist-jetty-config/src/main/resources/org/exist/jetty/etc/webapps/exist-webapp-context.xml#L5-L6), the [root controller](https://github.com/eXist-db/exist/blob/develop/exist-jetty-config/src/main/resources/webapp/WEB-INF/controller-config.xml#L45), and [the app's target collection](https://github.com/eXist-db/public-repo/blob/master/repo.xml#L10).

At present, the safest way to gather this data is from within the `controller.xql`, which uniquely has access to the `$exist:*` variables, but from there you can simply provide the needed value as a parameter to the views that need to construct a base-url.

File: index.html [hunk 3]
```diff
@@ -30,7 +30,7 @@ <h2>Packages</h2>
         <div class="col-md-4">
             <h2>Admin</h2>
             <p>
-                <a href="admin.html" class="btn btn-primary">
+                <a href="./admin" class="btn btn-primary">
```
Comment (joewiz) [inline index.html:33]: `./` can be dropped

File: controller.xql [hunk 4]
```diff
@@ -17,84 +17,93 @@ declare variable $app-root-absolute-url :=
...
+                <set-header name="Cache-Control" value="no-cache"/>
+            </forward>
+        </view>
+    </dispatch>
+
+(: legacy route :)
+else if (request:get-method() eq "GET" and $exist:path eq "/public/apps.xml") then
     <dispatch xmlns="http://exist.sourceforge.net/NS/exist">
-        <forward url="{$exist:controller}/modules/put-package.xq"/>
+        <forward url="/{$exist:controller}/modules/list.xq"/>
```
Comment (joewiz) [inline controller.xql:49]: Here and in the subsequent similar changes, the leading `/` added should be removed. `$exist:controller` already provides a leading `/`. (In this app, it's `/public-repo`.)

File: controller.xql [hunk 5]
```diff
@@ -105,37 +114,70 @@ else if (contains($exist:path, "/public/") and (ends-with($exist:resource, ".png
...
-    
-else
-    (: everything else is passed through :)
+
+(: redirect any unauthorized request to the login page :)
+else if (not(local:is-authorized-user()) and 
+    ($exist:path eq "/admin" or $exist:path eq "/publish")) then
     <dispatch xmlns="http://exist.sourceforge.net/NS/exist">
-        <cache-control cache="yes"/>
+        <forward url="login.html"/>
```
Comment (joewiz) [inline controller.xql:151]: > I favour absolute urls whenever possible.

... so let's include `{$exist:controller}` across the board ;)

File: controller.xql [hunk 6]
```diff
@@ -105,37 +114,70 @@ else if (contains($exist:path, "/public/") and (ends-with($exist:resource, ".png
...
+        </view>
+    </dispatch>
+
+(: 
+ : Protected resources:
+ : user is required to be logged in and member of a specific group 
+ :)
+else if (request:get-method() = ("GET", "POST") and $exist:path eq "/admin") then
+    <dispatch xmlns="http://exist.sourceforge.net/NS/exist">
+        <forward url="admin.html"/>
```
Comment (joewiz) [inline controller.xql:165]: Another opportunity for `{$exist:controller}` absolute URL consistency.

File: modules/config.xqm (hunk 1 above)
Comment (joewiz) [reply modules/config.xqm:16]: How about adding another variable here, rather than hardcoding `".user"` in those modules?

File: index.html (hunk 3 above)
Comment (joewiz) [reply index.html:33]: ... or is it needed for base-url resolution?

File: modules/log.xqm [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    format-date($date, "[Y]/[M01]")
+};
+
+declare function log:collection($date as xs:date) {
+    $config:logs-col || "/" || 
+    log:subcollection($date)
+};
+
+declare function log:document-name($date as xs:date) {
+    ``[public-repo-log-`{format-date($date, "[Y]-[M01]-[D01]")}`.xml]``
```
Comment (joewiz) [inline modules/log.xqm:42]: The original formulation in `config:log-doc-name` was more readable. Let's only resort to the string constructor to avoid escaping curly braces, etc.

**Comment by joewiz** [comment]
@line-o Really great stuff! Much improved routing and logging. Besides my inline comments, my only concern is:

I support the idea of dropping `.html` suffixes from URLs, but we have lots of existing links that point to them—especially the package info pages—from as recent as yesterday. Rather than return 404s for these, could we add routes that redirect them to the correct URLs? As with `modules/find.xql`, we should strive to support retired URLs where possible.

File: controller.xql [hunk 8]
```diff
@@ -105,37 +114,70 @@ else if (contains($exist:path, "/public/") and (ends-with($exist:resource, ".png
...
+        <view>
+            <forward url="{$exist:controller}/modules/view.xq">
+                <set-header name="Cache-Control" value="no-cache"/>
+            </forward>
+        </view>
+    </dispatch>
+
+else if (request:get-method() eq "POST" and $exist:path eq "/publish") then
+    <dispatch xmlns="http://exist.sourceforge.net/NS/exist">
+        <forward url="{$exist:controller}/modules/put-package.xq"/>
```
Comment (joewiz) [inline controller.xql:175]: How about renaming `put-package.xq` to `publish.xq`, to keep the names aligned?

File: index.html (hunk 3 above)
Comment (line-o) [reply index.html:33]: I wanted the href attributes to resemble one another `./`, `../`, `./admin` make it obvious that we are dealing with relative path that are resolved against something.

File: controller.xql (hunk 4 above)
Comment (line-o) [reply controller.xql:49]: OK, will do. I found one forward with leading slash and thought it's the way to go.

File: controller.xql (hunk 5 above)
Comment (line-o) [reply controller.xql:151]: good catch!

File: modules/log.xqm (hunk 7 above)
Comment (line-o) [reply modules/log.xqm:42]: I am so used to read template strings that it is actually the easier form to read.
What about a two line form with a intermediate variable:
```xquery
let $formatted-date := format-date($date, "[Y]-[M01]-[D01]")
return
  ``[public-repo-log-`{$formatted-date}`.xml]``
```

I can revert that, but I strongly disagree that string constructors should only be used to avoid escaping curly braces. Think of multiline strings and templates in general.
For me it is the other way around, I do use concatenation (`||`) only when a string constructor does not make sense.

File: controller.xql (hunk 8 above)
Comment (line-o) [reply controller.xql:175]: OK

**Comment by line-o** [comment]
I agree with the permanent redirects for index.html and admin.html

File: modules/log.xqm (hunk 7 above)
Comment (joewiz) [reply modules/log.xqm:42]: Are you going to change the whole app then or just this one?

**Comment by line-o** [comment]
@joewiz What do you mean? The URLs all have a canonical and one that will be redirected to. Or did I forget something?
On a side note: The controller seems to override or not set the status-code 301 but redirects with `302 FOUND` instead

**Comment by line-o** [comment]
Screenshot 2021-03-07 at 13 24 40

File: modules/config.xqm (hunk 2 above)
Comment (line-o) [reply modules/config.xqm]: What do you think about the current solution:
Use the base url constructed in the controller by adding as a parameter to all pages.
That way any further logic, like `is-proxied` can be implemented in a central location.

File: modules/log.xqm (hunk 7 above)
Comment (line-o) [reply modules/log.xqm:42]: No, I will revert it :)

File: controller.xql (hunk 8 above)
Comment (line-o) [reply controller.xql]: done

Comment (line-o) [reply controller.xql]: I ended up with `publish-package.xq`

File: modules/config.xqm (hunk 2 above)
Comment (joewiz) [reply modules/config.xqm]: Yes, I think the current solution is good. The parameter is added to all pages as needed.

File: index.html (hunk 3 above)
Comment (joewiz) [reply index.html:33]: Ok, thanks.

File: modules/log.xqm (hunk 7 above)
Comment (joewiz) [reply modules/log.xqm]: Thanks! (Phew!)

File: modules/app.xqm [hunk 9]
```diff
@@ -6,9 +6,10 @@ xquery version "3.1";
 
 module namespace app="http://exist-db.org/xquery/app";
 
+
+import module namespace templates="http://exist-db.org/xquery/templates";
```
Comment (joewiz) [inline modules/app.xqm:10]: They had been alphabetical by namespace prefix. If you want to group external module imports together, let's add a space between the groups.

Comment (joewiz) [reply modules/app.xqm:10]: I've applied this idea in a branch. Marking this as resolved in advance.

File: controller.xql [hunk 10]
```diff
@@ -17,84 +18,100 @@ declare variable $app-root-absolute-url :=
...
+        request:get-method() eq "GET" and
+        starts-with($exist:path, "/public/") and
+        (
+            ends-with($exist:resource, ".png") or
+            ends-with($exist:resource, ".svg")
+        )
+    ) then
     <dispatch xmlns="http://exist.sourceforge.net/NS/exist">
-        <forward url="{$exist:controller}/modules/get-icon.xq">
+        <forward url="/{$exist:controller}/modules/get-icon.xq">
```
Comment (line-o) [inline controller.xql:114]: we missed one leading slash :(

File: controller.xql (hunk 6 above)
Comment (line-o) [reply controller.xql:180]: and that is still here ! :(

File: modules/get-package.xq [hunk 11]
```diff
@@ -8,37 +8,37 @@ xquery version "3.1";
...
-            ( 
-                app:mkcol($config:logs-col, config:log-subcol($today)),
-                xmldb:store(config:log-col($today), config:log-doc-name($today), element public-repo-log { $event } )
-            )
     return
-        ()
+        log:event($event)
 };
 
+declare function local:log-package-not-found-event($filename as xs:string) as empty-sequence() {
```
Comment (line-o) [inline modules/get-package.xq:31]: I forgot to wire this in :)

File: packages.html [hunk 12]
```diff
@@ -1,10 +1,10 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<div data-template="templates:surround" data-template-with="templates/page-inner.html" data-template-at="container-body">
+<div data-template="templates:surround" data-template-with="templates/page.html" data-template-at="container-body">
     <div class="row">
         <div class="col-md-12">
             <h1>Public Application Repository</h1>
             <p>
-                <a href="../">Back to listing of all packages</a>
+                <a href="./">Back to listing of all packages</a>
```
Comment (line-o) [inline packages.html:7]: from `packages/something` the link back should still be `../` no?

File: modules/get-package.xq (hunk 11 above)
Comment (line-o) [reply modules/get-package.xq:31]: can be done at a later point

**Comment by line-o** [comment]
Ready to be merged.

File: packages.html (hunk 12 above)
Comment (joewiz) [reply packages.html:7]: In my testing of the app, this change (from `../` to `./`) works.

**Comment by joewiz** [comment]
Issues uncovered in this PR:

- [ ] As noted in https://github.com/eXist-db/public-repo/pull/62/commits/01a09a86abdfb95f68eb50a21066a532f6d5c310, the upload target in ant is broken.
- [ ] As noted in https://github.com/eXist-db/public-repo/pull/62#issuecomment-792303400, the controller overrides 301s set via response:set-status-code and responds with with 302 instead. Error handling should either use `<error-handler>` directives (not documented in, but modeled in https://github.com/HistoryAtState/hsg-shell/blob/master/controller.xql#L1737-L1751), or the app's URL rewriting/routing/error-handling needs should be migrated to roaster.

Opportunities for further enhancement:

- [ ] As noted in https://github.com/eXist-db/public-repo/pull/62/files#r588980150, we could refactor `$config:login-domain || ".user"` in controller.xql and publish-package.xq into a new config variable.
- [ ] As noted in https://github.com/eXist-db/public-repo/pull/62/files#diff-be7412746393b18ae23052e25cbb0ffb12fd8a007c659241956de4082ab1b775R62, we could log the eXist versions supplied by packageservice requests to apps.xml.
